tests/backup_restore_test.go

```diff
@@ -0,0 +1,284 @@
+	tx, err := db.Begin()
+	a.NoError(err)
+	defer tx.Rollback()
+	for i := 1; i <= numRecords; i++ {
+		_, err = tx.Exec(fmt.Sprintf("INSERT INTO %s VALUES (%d)", table, i))
+		a.NoError(err)
+	}
+	err = tx.Commit()
+	a.NoError(err)
+
+	// make a full backup
+	driver := getDbDriver(a, localhost, fmt.Sprintf("%d", port), username, database)
+	defer func() {
+		err := driver.Close(context.TODO())
+		a.NoError(err)
+	}()
+
+	buf := backup(a, driver, database)
+	// t.Logf("dump:\n%s", buf.String())
+
+	// drop all tables
+	_, err = db.Exec(fmt.Sprintf("DROP TABLE %s", table))
+	a.NoError(err)
+
+	// restore
+	err = driver.Restore(context.TODO(), bufio.NewScanner(buf), dbPlugin.RestoreConfig{})
+	a.NoError(err)
+
+	// validate data
+	rows, err := db.Query(fmt.Sprintf("SELECT * FROM %s ORDER BY id ASC", table))
+	a.NoError(err)
+	i := 1
+	for rows.Next() {
+		var col int
+		err := rows.Scan(&col)
+		a.NoError(err)
+		a.Equal(col, i)
+		i++
+	}
+	a.NoError(rows.Err())
+	a.Equal(numRecords+1, i)
+}
+
+// TestPITR tests the PITR behavior
+// The test plan is:
+// 0. prepare tables with foreign key constraints dependencies
+// 1. insert data and make a full backup (denoted as t0), which defines the checkpoint
+// 2. insert more data, and this is the point-in-time (denoted as t1) that we want to recover
+// 3. keep inserting data into the original tables
+// 4.1. set foreign_key_checks=OFF
+// 4.2. restore full backup at t0 to ghost tables
+// 4.3. apply binlog from t0 to t1 to ghost tables
+// 4.4. foreign_key_checks=ON
+// 5. lock tables and atomically swap original and ghost tables
+func TestPITR(t *testing.T) {
+	t.Parallel()
+	a := require.New(t)
+
+	localhost := "127.0.0.1"
+	port := getTestPort(t.Name())
+	username := "root"
+	database := "backup_restore"
+
+	insertData := func(db *sql.DB, a *require.Assertions, begin, end int) {
+		tx, err := db.Begin()
+		a.NoError(err)
+		defer tx.Rollback()
+
+		for i := begin; i < end; i++ {
+			_, err := tx.Exec(fmt.Sprintf("INSERT INTO t00 VALUES (%d)", i))
+			a.NoError(err)
+			_, err = tx.Exec(fmt.Sprintf("INSERT INTO t10 VALUES (%d, %d)", i, i))
+			a.NoError(err)
+			_, err = tx.Exec(fmt.Sprintf("INSERT INTO t11 VALUES (%d, %d)", i, i))
+			a.NoError(err)
+		}
+
+		err = tx.Commit()
+		a.NoError(err)
+	}
+
+	_, stop := mysql.SetupTestInstance(t, port)
+	defer stop()
+
+	db, err := sql.Open("mysql", fmt.Sprintf("root@tcp(localhost:%d)/mysql", port))
+	a.NoError(err)
+	defer db.Close()
+
+	_, err = db.Exec(fmt.Sprintf("CREATE DATABASE %s", database))
+	a.NoError(err)
+
+	_, err = db.Exec(fmt.Sprintf("USE %s", database))
+	a.NoError(err)
+
+	_, err = db.Exec(`
+	CREATE TABLE t00 (
+		id INT,
+		PRIMARY KEY (id),
+		CHECK (id > -1)
+	);
+	`)
+	a.NoError(err)
+	_, err = db.Exec(`
+	CREATE TABLE t10 (
+		id INT,
+		pid INT,
+		PRIMARY KEY (id),
+		UNIQUE INDEX (pid),
+		CONSTRAINT FOREIGN KEY (pid) REFERENCES t00(id) ON DELETE NO ACTION
+	);
+	`)
+	a.NoError(err)
+	_, err = db.Exec(`
+	CREATE TABLE t11 (
```

Comment:
what's idea behind creating another t11 which is the same as t10?

Reply:
This is a mistake, will delete t11.